### ISSUE TITLE
监控系统事件（网卡）

### ISSUE THREAD

**Issue by holdyounger** [holdyounger/Issue#55]
监控系统事件（网卡）

~~~cpp
#pragma once

#include &lt;netlistmgr.h&gt;
#include &lt;atomic&gt;

std::atomic_bool g_bNetWorkOnline(true);

class CNetworkListManagerEvent : public INetworkListManagerEvents
{
public:
	CNetworkListManagerEvent(std::function&lt;void(bool)&gt; netChangedHandler) 
		: m_ref(1)
		, netChangedHandler_(netChangedHandler)
	{

	}

	~CNetworkListManagerEvent()
	{

	}

	HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid, void** ppvObject)
	{
		HRESULT Result = S_OK;
		if (IsEqualIID(riid, IID_IUnknown))
		{
			*ppvObject = (IUnknown*)this;
		}
		else if (IsEqualIID(riid, IID_INetworkListManagerEvents))
		{
			*ppvObject = (INetworkListManagerEvents*)this;
		}
		else
		{
			Result = E_NOINTERFACE;
		}

		return Result;
	}

	ULONG STDMETHODCALLTYPE AddRef()
	{
		return (ULONG)InterlockedIncrement(&amp;m_ref);
	}

	ULONG STDMETHODCALLTYPE Release()
	{
		LONG Result = InterlockedDecrement(&amp;m_ref);
		if (Result == 0)
			delete this;
		return (ULONG)Result;
	}

	virtual HRESULT STDMETHODCALLTYPE ConnectivityChanged(
		/* [in] */ NLM_CONNECTIVITY newConnectivity)
	{
		if (newConnectivity == NLM_CONNECTIVITY_DISCONNECTED)
		{
			SP_DEV_LOGT(&quot;internet status changed. connect closed&quot;);

			// 网络断开
			netChangedHandler_(false);
		}
		else if ((newConnectivity &amp; NLM_CONNECTIVITY_IPV4_INTERNET) || (newConnectivity &amp; NLM_CONNECTIVITY_IPV6_INTERNET))
		{
			SP_DEV_LOGT(&quot;internet status changed. connect success&quot;);

			// 网络连接成功
			netChangedHandler_(true);
		}
		return S_OK;
	}

private:
	LONG m_ref;
	std::function&lt;void(bool)&gt; netChangedHandler_;
};


class CNetworkMonitor
{
public:
	CNetworkMonitor() {};
	~CNetworkMonitor() {};

private:
	// 监控网络状态变化
	static void InitNetworkMonitor()
	{
		SP_DEV_LOGT(&quot;InitNetworkMonitor&quot;);

		bool bSuccess = false;
		CoInitialize(NULL);
		//
		//  通过NLA接口获取网络状态
		//
		IUnknown* pUnknown = NULL;
		HRESULT Result = CoCreateInstance(CLSID_NetworkListManager, NULL, CLSCTX_ALL, IID_IUnknown, (void**)&amp;pUnknown);
		if (SUCCEEDED(Result))
		{
			INetworkListManager* pNetworkListManager = NULL;
			Result = pUnknown-&gt;QueryInterface(IID_INetworkListManager, (void**)&amp;pNetworkListManager);
			if (SUCCEEDED(Result))
			{
				VARIANT_BOOL IsConnect = VARIANT_FALSE;
				Result = pNetworkListManager-&gt;get_IsConnectedToInternet(&amp;IsConnect);
				if (SUCCEEDED(Result))
				{
					SP_DEV_LOGT(&quot;connect to internet:%s&quot;, IsConnect == VARIANT_TRUE ? &quot;TRUE&quot; : &quot;FALSE&quot;);
				}

				IConnectionPointContainer* pCPContainer = NULL;
				Result = pNetworkListManager-&gt;QueryInterface(IID_IConnectionPointContainer, (void**)&amp;pCPContainer);
				if (SUCCEEDED(Result))
				{
					IConnectionPoint* pConnectPoint = NULL;
					Result = pCPContainer-&gt;FindConnectionPoint(IID_INetworkListManagerEvents, &amp;pConnectPoint);
					if (SUCCEEDED(Result))
					{
						DWORD Cookie = NULL;

						CNetworkMonitor networkMonitor;
						std::function&lt;void(bool)&gt; monitorFunc = std::bind(&amp;CNetworkMonitor::NetworkStatusChangedCallback, networkMonitor, std::placeholders::_1);
						CNetworkListManagerEvent* pNetEvent = new(std::nothrow) CNetworkListManagerEvent(monitorFunc);
						if (pNetEvent)
						{
							Result = pConnectPoint-&gt;Advise((IUnknown*)pNetEvent, &amp;Cookie);
							if (SUCCEEDED(Result))
							{
								SP_DEV_LOGT(&quot;InitNetworkMonitor success&quot;);
								bSuccess = true;

								MSG msg;
								while (GetMessage(&amp;msg, NULL, 0, 0))
								{
									TranslateMessage(&amp;msg);
									DispatchMessage(&amp;msg);

									if (msg.message == WM_QUIT)
									{
										break;
									}
								}

								pConnectPoint-&gt;Unadvise(Cookie);
								pConnectPoint-&gt;Release();
							}
						}
					}
					pCPContainer-&gt;Release();
				}
				pNetworkListManager-&gt;Release();
			}
			pUnknown-&gt;Release();
		}
		CoUninitialize();
	}

	void NetworkStatusChangedCallback(bool status)
	{
		g_bNetWorkOnline = status;
		if (status)
		{
			if (g_sp_trust_model.get())
			{
				if (g_sp_trust_model-&gt;IsOnline())
				{
					//休眠唤醒后网络连接成功，自动发送一次心跳
					SP_LOGI(&quot;Sleep wakes up and automatically triggers a heartbeat&quot;);
					g_sp_trust_model-&gt;OnWakeToSendHeartBeat();

					SP_LOGI(&quot;rebuild trust nc&quot;);

					if (g_sp_trust_model-&gt;Session()-&gt;NeedsL3Tunnel() &amp;&amp; !g_sp_trust_model-&gt;Session()-&gt;IsTunnelSvrsEmpty())
					{
						// 网络连接成功后自动触发重连
						std::thread buildUtunThread([] {
							g_sp_trust_model-&gt;NotifyUpMessage(SP_TRUST_TOPIC_BUILD_UTUN, SPJSONObject());
							});
						buildUtunThread.detach();
					}
				}
				else
				{
					SP_LOGI(&quot;No login, will not rebuild trust nc&quot;);
				}
			}
			else
			{
				SP_LOGI(&quot;rebuild trust nc - g_sp_trust_model is null&quot;);
			}
		}
	}

public:
	void StartMonitor()
	{
		std::thread monitorThread(&amp;InitNetworkMonitor);
		monitorThread.detach();
	}
};
~~~

> blog link [监控系统事件（网卡）](https://holdyounger.github.io/OS/Windows/监控系统事件/)
